### PR TITLE
Allow from 0 to 5 preheat setups (bugfix)

### PR DESCRIPTION
**Pull request by GMagician**
- this will complete the "more than 2" preheats setup integration
- added possibility to have 0 or 1 preheats setup (freeing up memory)

Fix also #18450

### CODE REVIEW

File: Marlin/src/lcd/menu/menu_ubl.cpp [hunk 1]
```diff
@@ -315,37 +333,39 @@ void _lcd_ubl_invalidate() {
...
+                                          "M190 S" STRINGIFY(PREHEAT_##p##_TEMP_BED) "\n" \
+                                          "M109 S" STRINGIFY(PREHEAT_##p##_TEMP_HOTEND) "\n"  \
+                                          "G29 P1\n"  \
+                                          "M104 S0\n" \
+                                          "M140 S0" \
+                                        ))
+    #else
+      #define BUILD_MESH_GCODE_ITEM(p)  GCODES_ITEM(MSG_UBL_BUILD_MESH_M##p, PSTR(  \
+                                          "G28\n" \
+                                           "M109 S" STRINGIFY(PREHEAT_##p##_TEMP_HOTEND) "\n" \
```
Comment (thinkyhead) [inline Marlin/src/lcd/menu/menu_ubl.cpp:349]: All these preheat items are editable. One of the maintainers of UBL needs to catch up on that.

**Comment by thinkyhead** [comment]
Before this is merged, I suppose it would be good to get rid of the need to define so many preheat strings in the language files, and use substitution instead. The simplest substitution would be to use `S_FMT` which works with the `printf` functions. Alternatively, custom substitution with an extension to the menu items could also work, via `lcd_put_u8str_ind_P`.

The thing is, in future I want users to be able to choose the name of each preset from a larger list of 10 or more materials. So, these fixed language strings will need to be replaced to support that added flexibility anyway.

**Comment by GMagician** [comment]
> The simplest substitution would be to use S_FMT which works with the printf functions. Alternatively, custom substitution with an extension to the menu items could also work, via lcd_put_u8str_ind_P.

I'll check how to manage this.

**Comment by GMagician** [comment]
> I suppose it would be good to get rid of the need to define so many preheat strings

I checked how menu items work and nothing is ready to accept a dynamic string then I think some sort of new ACTION_ITEM must be created...something ACTION_ITEM_FMT? or completely wrong direction?

**Comment by thinkyhead** [comment]
> I'll check how to manage this.

There is currently a set of menu item macros which take an index (e.g., `MENU_ITEM_N_P`). This index value (`MenuItemBase::itemIndex`) is then passed on to `lcd_put_u8str_ind_P` to display a number with an optional "`E`" prefix. The `~` and `*` characters are used for these substitutions. I propose to add another substitution character "`$`" and another optional parameter taking a pointer to a list of PSTR pointers (which itself may be stored in PROGMEM). When `$` is seen, the index value from the menu item can be used as a lookup into the list of PSTR pointers, and the string from the list substituted in place.

So, it means extending `MenuItemBase` with a static pointer to a list of strings (or a general-use `void` pointer) to be used for this kind of extension. The pointer could be used in some instances for a single string substitution, or to point to some other data that the menu item will use for display.

File: Marlin/src/inc/Conditionals_post.h [hunk 2]
```diff
@@ -2101,11 +2101,9 @@
     #define PREHEAT_COUNT 3
   #elif defined(PREHEAT_2_LABEL)
     #define PREHEAT_COUNT 2
-  #else
+  #elif defined(PREHEAT_1_LABEL)
```
Comment (GMagician) [inline Marlin/src/inc/Conditionals_post.h:2104]: This should not be that way...lavel 1 check is already done above

Comment (GMagician) [reply Marlin/src/inc/Conditionals_post.h:2104]: ok I saw now...just remove on fisrt if. I'm trying to add same logic to all mesh preheats

**Comment by GMagician** [comment]
I'm honest, 'String list substitution' commit goes beyond my Marlin programming knowledge...

**Comment by thinkyhead** [comment]
I left the `MSG_PREHEAT_1` (etc.) language strings in place intentionally, because there is `ExtUI` (and other LCD) code which is far from ready to be adapted to incorporate this string substitution method.

**Comment by GMagician** [comment]
> I left the MSG_PREHEAT_1

then I'll revert and keep only configuration_post change

**Comment by thinkyhead** [comment]
Here are some of the materials that we are likely to include in the select list for the future when we can choose a material name for each preset…
```cpp
typedef enum MaterialType : uint8_t {
  MATERIAL_PLA,
  MATERIAL_PVA,
  MATERIAL_NYLON,
  MATERIAL_PVC,
  MATERIAL_TPU,
  MATERIAL_PVC,
  MATERIAL_CFPLA,
  MATERIAL_HTPLA,
  MATERIAL_PETG,
  MATERIAL_ABS,
  MATERIAL_ASA,
  MATERIAL_HIPS,
  MATERIAL_PC,
  MATERIAL_PEEK,
  //MATERIAL_CUSTOM1,
  //MATERIAL_CUSTOM2,
  //MATERIAL_CUSTOM3,
  //MATERIAL_CUSTOM4,
  MATERIAL_COUNT
} material_t;
```

**Comment by GMagician** [comment]
I was going to test this but I found that it can't be complied due to filament runout change (and not completed at time I made this).
~~Checking code I suspect now it works in reverse way. Il post a issue to get confirm~~

**Comment by GMagician** [comment]
I had to rebase it to compile and test. It seems not working it shows a $ instead of PLA/PETG, investigating

**Comment by thinkyhead** [comment]
I just did a test with the code in its current state, and it worked well for me. I think this can be merged for further ~~mayhem~~ testing and exercise.

File: Marlin/src/lcd/lcdprint.cpp [hunk 3]
```diff
@@ -57,10 +57,8 @@ lcd_uint_t lcd_put_u8str_ind_P(PGM_P const pstr, const int8_t ind, PGM_P const i
       if (n) n -= lcd_put_u8str_max_P((PGM_P)p, n * (MENU_FONT_WIDTH)) / (MENU_FONT_WIDTH);
       continue;
     }
-    else if (ch == '$') {
+    else if (ch == '$')
       n -= lcd_put_u8str_max_P(inStr, n * (MENU_FONT_WIDTH)) / (MENU_FONT_WIDTH);
-      continue;
```
Comment (GMagician) [inline Marlin/src/lcd/lcdprint.cpp]: That's not good, without continue it will print $ as well

**Comment by thisiskeithb** [comment]
@thinkyhead: The only issue I found was with 2004/character LCDs displaying a percent symbol after the right arrow (looks fine on a 12864/full graphic LCD). I dug through the related code and couldn't find an obvious reason for it:

image

Fixed in https://github.com/MarlinFirmware/Marlin/pull/18644

**Comment by GMagician** [comment]
it seems writing 1 char less. In my display '>' is on last column

**Comment by thisiskeithb** [comment]
> it seems writing 1 char less. In my display '>' is on last column

Which LCD do you have? It looks fine with a 12864/full graphic controller:
image

**Comment by thisiskeithb** [comment]
It repeats the last character from the menu item above it on the 2004 LCDs. Here's a photo from the Configuration menu:
image

Fixed in #18644

**Comment by GMagician** [comment]
> Which LCD do you have? It looks fine with a 12864/full graphic controller:

Yes this is what I have

**Comment by thisiskeithb** [comment]
> Yes this is what I have

I dug through the code, but couldn't find an obvious reason for the error on 2004/character LCDs. Have any ideas?

**Comment by GMagician** [comment]
Something has been changed on the for cycle, but this shouldn't "repeat" last char of previous line

**Comment by thinkyhead** [comment]
> Something has been changed on the for cycle, but this shouldn't "repeat" last char of previous line

It's just not writing over it as it scrolls (or writing over the contents of the previous screen).

**Pull request merged**